### PR TITLE
feat(tiptap): Emoji Support

### PR DESCRIPTION
**Pull request by jmtaber129**
# Description

Fixes https://github.com/ParabolInc/parabol/issues/6883

Add emoji support by typing `:`.  Reuse the emoji menu used in our Draft.js editor to show emojis in TipTap.

Changes:
* Refactor existing emoji menu to decouple the menu from the draft.js editor
* Add a new tiptap emoji component that uses the [Suggestion](https://github.com/ueberdosis/tiptap/tree/main/packages/suggestion) tiptap utility to render the emoji menu.

This uses a pattern used by many TipTap React components in which the component is rendered by simply passing in the editor - no added configuration in the config (see the [React Bubble Menu](https://github.com/ueberdosis/tiptap/blob/9a254bf9daf6d839bd02c968e14837098b903b38/packages/react/src/BubbleMenu.tsx)).

**note**: one test case this fails is when the cursor is part-way through the emoji query - the emoji suggestions are only based on the query from `:` to the cursor, and selecting an emoji will only replace the text from `:` to the cursor.  E.g. if I type out `:joy`, and place my cursor between `o` and `y` and click the `:joy:` emoji, the resultant text will look like `😂y`.  **This is an issue with the tiptap `Suggestion` utility** (see the example on their [Emojis package docs](https://tiptap.dev/api/nodes/emoji)), so we'll need to either PR a fix into that package, or do some custom range selection to fix this. This likely isn't enough of a degradation to prioritize fixing in the near-term, but we should stay attentive to feedback about this.

## Demo
https://www.loom.com/share/fa5cb5c6b84445e9bbbdf877df8c926b

## Testing scenarios
Do the following on both standups and Draft.js:
* Type out part of an emoji by typing `:<emoji`, then click an emoji
* Type out part of an emoji, select elsewhere in the textbox, then select the end of the emoji query, and click an emoji
* Type out part of an emoji, click out of the textbox to lose focus, then select the end of the emoji query, and click an emoji
* Type out part of an emoji, then type a space and see that the menu disappears.

## Final checklist

- [X] I checked the [code review guidelines](../docs/codeReview.md)
- [X] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [X] I have performed a self-review of my code, the same way I'd do it for any other team member
- [X] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [X] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [X] I added the label `One Review Required` if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [X] PR title is human readable and could be used in changelog

### CODE REVIEW

File: packages/client/components/TaskEditor/withEmojis.tsx [hunk 1]
```diff
@@ -29,10 +31,22 @@ const withEmojis = (ComposedComponent) => {
...
-      e.preventDefault()
+    static getDerivedStateFromProps(nextProps, prevState) {
+      const {editorState} = nextProps
+      return {
+        // clicking on a menu will cause the editorStateSelection to lose focus, so we persist the last state before that point
+        focusedEditorState: editorState.getSelection().getHasFocus()
+          ? editorState
+          : prevState.focusedEditorState
+      }
+    }
```
Comment (jmtaber129) [inline packages/client/components/TaskEditor/withEmojis.tsx:42]: I'm not sure if this exact logic is necessary, but I've copied this from `withEmojis` to maintain identical behavior

File: packages/client/components/promptResponse/EmojiMenuTipTap.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  const [emojiQuery, setEmojiQuery] = useState('')
+  const [range, setRange] = useState<Range | null>(null)
+
+  useEffect(() => {
+    if (tiptapEditor.isDestroyed) {
+      return
+    }
+
+    const plugin = Suggestion({
+      pluginKey: pluginKey as any,
```
Comment (BartoszJarocki) [inline packages/client/components/promptResponse/EmojiMenuTipTap.tsx:26]: +1 just curious, why do we need to cast here?

**Review by BartoszJarocki — APPROVED** [review]
LGTM! the only improvement I'd love to see is being able to navigate the emoji menu with the keyboard (just like on slack). right now, when I want to select the emoji from the menu, I have to use a mouse, which isn't perfect.

File: packages/client/components/promptResponse/EmojiMenuTipTap.tsx (hunk 2 above)
Comment (jmtaber129) [reply packages/client/components/promptResponse/EmojiMenuTipTap.tsx:26]: The types between the `PluginKey` I'm using and the `PluginKey` `Suggestion` expects don't quite match up exactly, so we need to cast to work around that.

Added a comment + TODO for this.

**Comment by jmtaber129** [comment]
> I'd love to see is being able to navigate the emoji menu with the keyboard

This would be awesome! Our existing emoji menu on draft.js doesn't seem to support this, but that would be a nice enhancement to take on later.

**Comment by BartoszJarocki** [comment]
yeah, I didn't mean this PR, I'll create a task for it. nice work!

**Pull request merged**